client/web/src/search/useSourcegraphCodeIntelligence.ts

```diff
@@ -0,0 +1,71 @@
+import * as Monaco from 'monaco-editor'
+import { useEffect } from 'react'
+import { Observable } from 'rxjs'
+
+import { SearchPatternType } from '@sourcegraph/shared/src/graphql/schema'
+import { getDiagnostics } from '@sourcegraph/shared/src/search/query/diagnostics'
+import { getProviders } from '@sourcegraph/shared/src/search/query/providers'
+import { scanSearchQuery } from '@sourcegraph/shared/src/search/query/scanner'
+import { SearchSuggestion } from '@sourcegraph/shared/src/search/suggestions'
+
+export const SOURCEGRAPH_SEARCH = 'sourcegraphSearch' as const
+
+/**
+ * Adds code intelligence for the Sourcegraph search syntax to Monaco.
+ */
+export function useSourcegraphSearchCodeIntelligence(
+    fetchSuggestions: (query: string) => Observable<SearchSuggestion[]>,
+    options: {
+        patternType: SearchPatternType
+        globbing: boolean
+        interpretComments?: boolean
+        isSourcegraphDotCom?: boolean
+    }
+): void {
+    useEffect(() => {
+        // Register language ID
+        Monaco.languages.register({ id: SOURCEGRAPH_SEARCH })
+
+        // Register providers
+        const providers = getProviders(fetchSuggestions, options)
+        const setTokensProviderDisposable = Monaco.languages.setTokensProvider(SOURCEGRAPH_SEARCH, providers.tokens)
+        const registerHoverProviderDisposable = Monaco.languages.registerHoverProvider(
+            SOURCEGRAPH_SEARCH,
+            providers.hover
+        )
+        const registerCompletionItemProviderDisposable = Monaco.languages.registerCompletionItemProvider(
+            SOURCEGRAPH_SEARCH,
+            providers.completion
+        )
+
+        return () => {
+            setTokensProviderDisposable.dispose()
+            registerHoverProviderDisposable.dispose()
+            registerCompletionItemProviderDisposable.dispose()
+        }
+    }, [fetchSuggestions, options])
+}
+
+export function useSourcegraphSearchDiagnostics(
```

Comment:
```suggestion
export function useQueryDiagnostics(
```

as above :-) again feel free to punt and do separately.